rpc,server: port server interceptors to drpc

shaikzakiriitm opened the pull request:
As part of migration from grpc to drpc, we need to migrate all the server interceptors within the cockroachdb to their drpc variants.
Here, we move multiple interceptors 1)Stopper interceptor, 2)Metrics recorder, 3)Access control filtering interceptor. Existing metrics recorder logic is refactored with factory methods to generate two variants (grpc, drpc) of metrics interceptor.

**Note**: This pr uses the changes from https://github.com/cockroachdb/drpc/pull/7

Epic: CRDB-49357
Fixes: #144373
Part of: CRDB-49359

Release note: None

shaikzakiriitm committed:
rpc,server: port server interceptors to drpc

As part of migration from grpc to drpc, we need to migrate
all the server interceptors within the cockroachdb to their drpc
variants.
Here, we move multiple interceptors 1)Stopper interceptor,
2)Metrics recorder, 3)Access control filtering interceptor.
Existing metrics recorder logic is refactored with factory methods
to generate two variants (grpc, drpc) of metrics interceptor.

Epic: CRDB-49357
Fixes: #144373
Part of: CRDB-49359

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/148256)
<!-- Reviewable:end -->

shaikzakiriitm committed:
rpc,server,tracing: tracing server interceptor in drpc

As part of drpc migration, server side tracing interceptors
are ported to drpc as part of this change.
Common logic used in both grpc/drpc tracing server interceptors
are moved to tracing/utils.go for reusability.
Drpc tracing interceptor is injected into the drpc server.

Epic: CRDB-51562
Fixes: #148350
Part of: CRDB-49359

Release note: None

shaikzakiriitm committed:
rpc,server: auth server interceptor in drpc

<what was there before: Previously, ...>
Adding authentication and authorization server interceptor for
drpc server.

Epic: CRDB-49353
Fixes: #144369
Part of: CRDB-49359

Release note: None
